File: backend/app/schemas/snapshot.py
"""快照管理相关的Schema定义"""
from datetime import date
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class SnapshotGenerateRequest(BaseModel):
    """单日快照生成请求"""
    portfolio_code: str
    target_date: date


class SnapshotRecalculateRequest(BaseModel):
    """区间快照重算请求"""
    portfolio_code: Optional[str] = None  # None表示所有活跃组合
    start_date: date
    end_date: date


class SnapshotCatchUpRequest(BaseModel):
    """快照追平请求（issue #84）"""
    portfolio_code: str
    to_date: date


class SnapshotGenerateNextRequest(BaseModel):
    """生成下一交易日快照请求（issue #84）"""
    portfolio_code: str


class SnapshotCatchUpResult(BaseModel):
    """快照追平结果（issue #84）

    逐日 checkpoint 语义：失败时 generated_dates 中已成功的日子已落库，
    failed_date/error 标记中断点。
    """
    portfolio_code: str
    to_date: str  # ISO格式
    generated_count: int
    generated_dates: List[str]  # ISO格式日期列表（升序）
    latest_snapshot_date: Optional[str] = None
    message: Optional[str] = None
    failed_date: Optional[str] = None
    error: Optional[str] = None
    # #305：逐日告警（带 date 键）与自动确认结果透传，无则 None
    warnings: Optional[List[Dict[str, Any]]] = None
    auto_confirmed: Optional[List[Dict[str, Any]]] = None
    # #305：中断点的结构化错误（error 仍为消息文本，向后兼容）
    error_code: Optional[str] = None
    error_details: Optional[Dict[str, Any]] = None


class SnapshotGenerateNextResult(BaseModel):
    """生成下一交易日快照结果（issue #84）"""
    success: bool
    message: str
    portfolio_code: str
    generated_date: str  # ISO格式
    total_value: Optional[float] = None
    total_shares: Optional[float] = None
    unit_price: Optional[float] = None
    warnings: Optional[List[Dict[str, Any]]] = None
    # #305：自动确认结果透传（含失败条目），无则 None
    auto_confirmed: Optional[List[Dict[str, Any]]] = None


class ValidationCheckResult(BaseModel):
    """数据校验结果"""
    check_type: str  # "trading_day", "pending_transactions", "price_data", "share_change_events"
    status: str  # "passed", "failed", "warning"
    message: str


class SnapshotValidationResult(BaseModel):
    """快照依赖数据校验结果"""
    portfolio_code: str
    target_date: date
    is_valid: bool
    checks: List[ValidationCheckResult]


class SnapshotGenerationResult(BaseModel):
    """快照生成结果"""
    success: bool
    message: str
    portfolio_code: str
    snapshot_date: date
    total_value: Optional[float] = None
    total_shares: Optional[float] = None
    unit_price: Optional[float] = None
    # 非阻断性告警（如 event_zeroed_position；无告警时为 None，旧客户端可忽略）
    warnings: Optional[List[Dict[str, Any]]] = None


class RecalculationPortfolioResult(BaseModel):
    """单个组合的重算结果"""
    portfolio_code: str
    processed_dates: List[str]  # ISO格式日期列表
    total_processed: int
    errors: List[Dict[str, Any]]
    # 逐日重建累积的告警（如 event_zeroed_position；负现金已走 NEGATIVE_CASH 进 errors）
    warnings: Optional[List[Dict[str, Any]]] = None
    # #305：逐日自动确认结果透传（含 auto_confirm_failed 条目），空列表为 []
    auto_confirmed: Optional[List[Dict[str, Any]]] = None
    # #305：删旧快照级联回退的申购记录，无则 []
    cascaded_unconfirmed: Optional[List[Dict[str, Any]]] = None
    # #305：end_date 因后续快照自动扩展到的实际重算终点（未扩展则无此字段）
    end_date_extended_to: Optional[str] = None


class RecalculationResult(BaseModel):
    """区间重算结果"""
    success: bool
    message: str
    results: List[RecalculationPortfolioResult]


class SnapshotListItem(BaseModel):
    """快照历史列表项（#146）。涨跌不由后端给——前端按相邻行 unit_price 推导。"""
    snapshot_date: date
    unit_price: float
    total_shares: float
    total_value: float
    in_transit_total: float

    class Config:
        from_attributes = True


class SnapshotListResponse(BaseModel):
    """快照历史列表响应。total = 过滤后全量计数（limit 截断前），防无声截断。"""
    portfolio_code: str
    items: List[SnapshotListItem]
    total: int
    limit: int


class SnapshotStatusResponse(BaseModel):
    """组合快照状态响应"""
    portfolio_code: str
    latest_snapshot_date: Optional[date] = None
    total_snapshots: int
    first_snapshot_date: Optional[date] = None
    missing_dates: List[str]  # 缺失的交易日（ISO格式）
    # issue #71：最新快照日 CASH 持仓 cash_amount < 0 的平台清单（正常为空）
    negative_cash_platforms: List[str] = []
    # issue #156：组合自动快照开关（仅约束自动任务，手动生成/重算不受影响）
    auto_snapshot_enabled: bool = False

File: frontend/src/types/snapshot.ts
export interface SnapshotValidationCheck {
  check_type: string;
  status: "passed" | "failed" | "warning";
  message: string;
}

export interface SnapshotValidationResult {
  portfolio_code: string;
  target_date: string;
  is_valid: boolean;
  checks: SnapshotValidationCheck[];
}

// 后端非阻断告警（issue #71：如负现金）
export type SnapshotWarning = Record<string, unknown>;

// auto_confirm_after_snapshot 逐条结果（#305：成功/失败均透传，失败携 code）
export type AutoConfirmEntry = Record<string, unknown>;

export interface SnapshotGenerationResult {
  success: boolean;
  message: string;
  portfolio_code: string;
  snapshot_date: string;
  total_value?: number;
  total_shares?: number;
  unit_price?: number;
  warnings?: SnapshotWarning[] | null;
}

export interface RecalculationPortfolioResult {
  portfolio_code: string;
  processed_dates: string[];
  total_processed: number;
  errors: Array<{
    date: string;
    error: string;
    code?: string;
    details?: Record<string, unknown>;
  }>;
  warnings?: SnapshotWarning[] | null;
  /** #305：逐日 auto_confirm 结果（含 auto_confirm_failed 条目与 code） */
  auto_confirmed?: AutoConfirmEntry[] | null;
  /** #305：删旧快照级联回退的申赎明细 */
  cascaded_unconfirmed?: AutoConfirmEntry[] | null;
  /** #305：end_date 后存在快照时自动扩展到的实际重算终点 */
  end_date_extended_to?: string | null;
}

export interface RecalculationResult {
  success: boolean;
  message: string;
  results: RecalculationPortfolioResult[];
}

export interface SnapshotStatusResponse {
  portfolio_code: string;
  latest_snapshot_date?: string;
  total_snapshots: number;
  first_snapshot_date?: string;
  missing_dates: string[];
  /** 最新快照日 CASH 持仓 cash_amount < 0 的平台清单（正常为空） */
  negative_cash_platforms?: string[];
  /** 组合级自动快照开关（issue #156），随组合配置返回 */
  auto_snapshot_enabled: boolean;
}

// ---- #146 快照管理页重构 ----

export interface SnapshotListItem {
  snapshot_date: string;
  unit_price: number;
  total_shares: number;
  total_value: number;
  in_transit_total: number;
}

export interface SnapshotListResponse {
  portfolio_code: string;
  items: SnapshotListItem[];
  /** 过滤后全量计数（limit 截断前），total > items.length 即被截断 */
  total: number;
  limit: number;
}

export interface SnapshotCatchUpResult {
  portfolio_code: string;
  to_date: string;
  generated_count: number;
  generated_dates: string[];
  latest_snapshot_date?: string | null;
  message?: string | null;
  failed_date?: string | null;
  error?: string | null;
  /** #305：中断错误的结构化 code/details（error 仍为消息文本） */
  error_code?: string | null;
  error_details?: Record<string, unknown> | null;
  /** #305：逐日生成累积的非阻断告警（每条带 date） */
  warnings?: SnapshotWarning[] | null;
  auto_confirmed?: AutoConfirmEntry[] | null;
}

export interface SnapshotGenerateNextResult {
  success: boolean;
  message: string;
  portfolio_code: string;
  generated_date: string;
  total_value?: number;
  total_shares?: number;
  unit_price?: number;
  warnings?: SnapshotWarning[] | null;
  auto_confirmed?: AutoConfirmEntry[] | null;
}

export interface RecalculateAsyncSubmitResult {
  job_id: number;
  status: string;
  message: string;
}

export interface BulkDeleteDryRunResult {
  dry_run: true;
  portfolio_code: string;
  from_date: string;
  count: number;
  /** 将删除的快照日期（倒序） */
  snapshot_dates: string[];
}

export interface BulkDeleteResult {
  success: boolean;
  message: string;
  deleted_count: number;
  details?: Array<{
    snapshot_date: string;
    deleted: number;
    cascaded_subs: number;
    cascaded_events: number;
  }>;
}
